Add address autocomplete component

opr opened the pull request:
### Changes proposed in this Pull Request:

This PR is based on #60945 

- Create a new component to replace the `ValidatedTextInput` for address_1
- This new component wraps `ValidatedTextInput` but also has space for a suggestions component if needed.
- Subscribes to the cart data store to handle updating the current provider when country changes.

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install 
[mock-address-provider.zip](https://github.com/user-attachments/files/22429171/mock-address-provider.zip) and
[mock-address-provider-germany.zip](https://github.com/user-attachments/files/22426528/mock-address-provider-germany.zip)

2. Go to WC -> Settings -> General and check `Enable predictive address search`. Set `Mock address provider Germany` as your preferred provider.
3. Go to the Checkout block
4. In the shipping address, set your country to Germany, in the billing address, set it to US.
5. In the console, type `wc.addressAutocomplete.activeProvider` and ensure the billing provider is `mock-provider` and the shipping provider is `mock-provider-germany`.
6. Switch shipping to USA and repeat step 5. Both providers should be `mock-provider`.
7. Try to check out with an empty the address_1 field and ensure you're prevented from doing so by validation errors.
8. Fill address_1 and check out. Ensure the data is saved correctly for shipping and billing.

#### Test in shortcode checkout too

- Since this PR removes the script dependency on `woocommerce` and `jquery` please check that address autocomplete still works in the shortcode checkout.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

opr committed:
Add address providers to asset data registry

opr committed:
Add address autocomplete component

opr committed:
Add styling for address autocomplete input

opr committed:
Replace address1 input with autocomplete input

opr committed:
Remove unneeded ref

opr committed:
Remove unused var

opr committed:
Don't run autocomplete code when in blocks context

opr committed:
Don't depend on woocommerce or jquery

opr committed:
Remove unused imports and duplicate type def

opr committed:
Allow active provider to be null TS

opr committed:
Add changelog

opr committed:
Add unit test to cover autocomplete component first iteration

opr committed:
Fix lint issues in test

opr committed:
Don't output providers if option is disabled

tjcafferkey commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/address-autocomplete/address-autocomplete.tsx:
```diff
@@ -0,0 +1,122 @@
+		// Set globally as this is going to be the source of truth where the actual provider objects are stored.
+		if ( window?.wc?.addressAutocomplete?.activeProvider ) {
+			window.wc.addressAutocomplete.activeProvider[ addressType ] = null;
+		}
```
I think if we're resetting values on the window object so they're in sync with the data store we should handle this at a lower level such as the reducer. Otherwise there is a risk this becomes out of sync during refactors or future iterations etc.

opr replied:
Good idea, I moved this to a `subscribe` callback on the cart data store. When the country changes it will check and reset the active provider.

opr committed:
Handle address changes in data store rather than component

opr committed:
Update params docs and export function for testing

opr committed:
Replace tests for autocomplete component

opr committed:
Export subscription function for easy testing

opr committed:
Add tests for autocomplete address subscription

opr committed:
Harden server providers check for cart test

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/address-autocomplete/address-autocomplete.tsx:
```diff
@@ -0,0 +1,35 @@
+			<ValidatedTextInput
+				{ ...props }
+				id={ id }
+				onChange={ props.onChange }
```
`onChange` is already in `...props` above. 
```suggestion

```

opr replied:
Nice, thanks!

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/data/cart/index.ts:
```diff
@@ -77,6 +84,113 @@ window.addEventListener( 'load', () => {
 // Pushes changes whenever the store is updated.
 subscribe( pushChanges, store );
 
+// Update address providers whenever the country changes.
```
I recommend creating a new data store for the autocomplete logic. 
- we separate concerns & decoupling, the cart store should not be handling autocomplete logic
- the new store will have one responsibility, making it easier to extend, decouple, etc
- we should be calculating `previousShippingCountry` and `previousBillingCountry` only after cart actions that impact the address

opr replied:
> I recommend creating a new data store for the autocomplete logic.

I agree that it's a separation of concerns, but my worry is the overhead and maintenance burden added by creating a new store to track two values. I also don't see the data store expanding beyond this. Personally I see checkout UI as part of the checkout data store, for example `orderNotes` is in there, so is `prefersCollection`, number of ongoing calculations (e.g. applying coupons) so keeping it in `wc/store/checkout` is justified imo.

> the new store will have one responsibility, making it easier to extend, decouple, etc

I can't disagree with this, but the data store is a proxy for the underlying mechanism which is stored on `window.wc.addressAutocomplete`. Technically, we don't need to use data stores at all, we could just use the window object, but I added the data store aspect to make the code more idiomatic, readable, and familiar to blocks developers.

> we should be calculating previousShippingCountry and previousBillingCountry only after cart actions that impact the address

This isn't possible with `subscribe` the callback is executed on any action to the data store. I think the immediate check for address changes is the earliest we can do this. Country changes can come from multiple different places btw: setting shipping or billing address, `receiveCart`, or `extensionCartUpdate`. I think making this logic run on those routes would make the code more complex for what would only be (if at all) a minor performance increase.

opr replied:
> I still need to look at plugins/woocommerce/client/legacy/js/frontend/address-autocomplete.js. I'm curious why did you choose to handle the logic for the both areas in the same file?

The only logic shared here is the registration logic, the rest of the file doesn't actually execute in a blocks context. This could be decoupled into a separate file to reduce the amount of JS loaded onto checkout, I'll look into if that's possible, or if it's something we address as a followup.

ralucaStan replied:
> Personally I see checkout UI as part of the checkout data store, for example orderNotes is in there, so is prefersCollection, number of ongoing calculations (e.g. applying coupons) so keeping it in wc/store/checkout is justified imo.

In my opinion, the autocomplete is not just UI, it's a complex feature that could grow further. But for now, as we discussed, because we're only storing these 2 values, we can keep it in the checkout store, as creating a new store will make it public API. But we should be moving to a separate store once this feature grows. 

My concern was also about the impact on the cart store, which was bloated by the autocomplete logic, so thank you for the changes in https://github.com/woocommerce/woocommerce/pull/60955/commits/7444b9e4eb03256f0cb37d160d68b853b3071f05

ralucaStan reviewed (changes requested):
I left a partial review, I'll resume in a couple of hours. 

I still need to look at `plugins/woocommerce/client/legacy/js/frontend/address-autocomplete.js`. I'm curious why did you choose to handle the logic for the both areas in the same file? 

Without looking into it, I'm thinking we could use an autocomplete data store to store providers for the checkout block, and keep the logic for legacy as is. We could reuse parts of it, but providers information could be accesed differently depending on the checkout type.

opr committed:
Output data on "common" script

opr committed:
Separate autocomplete registration into its own script

opr committed:
Update tests to work with new shared structure

opr committed:
Dequeue shortcode autocomplete scripts in checkout block

opr committed:
Don't run subscription if no providers are registered

opr committed:
Fix JS import path

opr commented:
@ralucaStan I've updated the code now, the following changes are applied:

- Make a "common" script that handles registration and tracking the providers. This means we don't need to duplicate code betweeen shortcode and checkout.
- Only enqueue the "shortcode" JS when checkout block is not in use.
- Prevent running the subscription callback if no providers were registered client side.

I also updated [mock-address-provider.zip](https://github.com/user-attachments/files/22429171/mock-address-provider.zip)

opr committed:
Update import path for legacy JS common module

opr committed:
Move autocomplete subscription into its own file

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/data/cart/index.ts:
```diff
@@ -77,6 +84,117 @@ window.addEventListener( 'load', () => {
+);
+
+export const autocompleteSubscription = () => {
+	if ( ! window?.wc?.addressAutocomplete?.providers ) {
```
👏

opr committed:
Update tests to reflect move to checkout data store

opr committed:
Update plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/address-autocomplete/address-autocomplete.tsx

Co-authored-by: Raluca Stan <ralucastn@gmail.com>

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/data/cart/index.ts:
```diff
@@ -9,6 +9,12 @@ import {
 	select,
 } from '@wordpress/data';
 import { controls as dataControls } from '@wordpress/data-controls';
+import { getSettingWithCoercion } from '@woocommerce/settings';
```
you can remove these and the CheckoutStoreDescriptor type below

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/data/checkout/index.ts:
```diff
@@ -25,4 +26,7 @@ export type CheckoutStoreDescriptor = typeof store;
 
 subscribe( pushChanges, store );
 
+// Subscribe to country changes to update address autocomplete provider
```
It would be good to add a comment here that this approach is needed because there are a lot of actions that modify the addresses.

opr committed:
Lint fixes

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/data/checkout/subscriptions/autocomplete.ts:
```diff
@@ -0,0 +1,121 @@
+	}
+};
+
+export const subscribeToCountryChanges = () => {
```
could we rename this to `subscribeToCartChanges`? It reflects the behaviour better as we're not only targeting address or country changes.

ralucaStan replied:
Actually, because the subscription runs when coupons are applied, or when phone inputs change, I was thinking what do you think if, outside of this PR, we could create 2 new actions, for when the billing and the shipping countries change? We'd have to hook in multiple areas, but we'd have a more deterministic approach.

opr replied:
I agree it's probably the correct approach. I do want to tackle it in a follow up though, making these changes here is going to expand the scope too much. I have tested it out locally and it's doable, I just don't want to inflate the diff.

ralucaStan commented on plugins/woocommerce/client/legacy/js/frontend/address-autocomplete.js:
```diff
@@ -1,992 +1,969 @@
  *
- * @param {Object} provider The provider object
- * @return {boolean} Whether the registration was successful
+ * Note: The core registration logic and provider management is now handled
```
```suggestion
 * Note: The core registration logic and provider management is handled
```

ralucaStan reviewed (changes requested):
Thank you for your work on this PR, I think it brings a good foundation for tautocomplete in the blocks area.

I've left some small comments, as you already addressed the initial comments I had.

opr committed:
Rename subscription function

opr committed:
Fix typo

opr committed:
Create useUpdatePreferredAutocompleteProvider hook

opr committed:
Make AddressAutocomplete use new hook

opr committed:
Update address form types to use shipping/billing

opr committed:
Remove subscription to data store to update providers

opr committed:
Show ValidatedTextInput only if no providers registered

opr committed:
Update test to run against new hook

opr committed:
Improve length check

opr committed:
Add tests for AdressAutocomplete component

opr committed:
Only output addressType if it is the autocomplete component

ralucaStan reviewed (approved):
I've retested this PR and looked at the latest changes.
I've tested with 1 missing provider, and it didn't break. The checkout store has the correct providers.
The resources are loaded as expected for the different checkouts (block, legacy) 

Great work @opr! You really nailed the implementation of the hook, the separation of the resources & providers for the legacy and block checkout. All worth it IMO!

Let's merge!
